File: generator/src/test/java/ch/bailu/gtk/TestTypeModel.java
package ch.bailu.gtk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

import ch.bailu.gtk.converter.NamespaceType;
import ch.bailu.gtk.model.type.CType;
import ch.bailu.gtk.model.type.ClassType;
import ch.bailu.gtk.model.type.JavaType;
import ch.bailu.gtk.table.StructureTable;

public class TestTypeModel {


	@Test
	void testCType() {
		CType ctype = new CType("void");

		assertFalse(ctype.isConst());
		assertFalse(ctype.isSinglePointer());
		assertEquals("void", ctype.getType());
		assertFalse(ctype.contains("ApplicationFlags"));


		ctype = new CType("GApplicationFlags");

		assertFalse(ctype.isConst());
		assertFalse(ctype.isSinglePointer());
		assertTrue(ctype.isDirectType());
		assertEquals("GApplicationFlags", ctype.getType());
		assertTrue(ctype.contains("ApplicationFlags"));


		ctype = new CType("PangoAttrShape*");

		assertFalse(ctype.isConst());
		assertTrue(ctype.isSinglePointer());
		assertFalse(ctype.isDirectType());
		assertEquals("PangoAttrShape*", ctype.getType());


		ctype = new CType("const gchar*");
		assertTrue(ctype.isConst());
		assertTrue(ctype.isSinglePointer());


		ctype = new CType("gchar**");
		assertFalse(ctype.isConst());
		assertFalse(ctype.isSinglePointer());

	}


	@Test
	void testJavaType() {
		JavaType jtype = new JavaType("int");

		assertEquals("int", jtype.getApiTypeName());
		assertTrue(jtype.isValid());


		jtype = new JavaType("void");
		assertEquals("void", jtype.getApiTypeName());
		assertTrue(jtype.isValid());
		assertTrue(jtype.isVoid());


		jtype = new JavaType("void");
		assertEquals("void", jtype.getApiTypeName());
		assertTrue(jtype.isValid());
		assertTrue(jtype.isVoid());

		jtype = new JavaType("guint8");
		assertEquals("int", jtype.getApiTypeName());
		assertTrue(jtype.isValid());
		assertFalse(jtype.isVoid());
	}

	@Test
	void testNamespaceType() {
		NamespaceType type = new NamespaceType("namespace", "namespace.Application");

		assertTrue(type.isValid());
		assertEquals("namespace", type.getNamespace());
		assertEquals("Application", type.getName());


		type = new NamespaceType("namespace", "Application");

		assertTrue(type.isValid());
		assertEquals("namespace", type.getNamespace());
		assertEquals("Application", type.getName());



		type = new NamespaceType("namespace2", "namespace.Application");

		assertTrue(type.isValid());
		assertEquals("namespace", type.getNamespace());
		assertEquals("Application", type.getName());


		NamespaceType ntype = new NamespaceType("GObject", "ParamSpecInt64");
		assertEquals("ParamSpecInt64", ntype.getName());
		assertEquals("gobject", ntype.getNamespace());
		assertTrue(ntype.isValid());
	}


	@Test
	void testClassType() {
		ClassType classType = new ClassType("namespace", "Application", "GtkApplication", false);
		assertFalse(classType.isClass());
		assertFalse(classType.isDirectType());

		classType = new ClassType("namespace", "Application", "GtkApplication", true);
		assertFalse(classType.isClass());


		StructureTable.INSTANCE.add("namespace", "Application");
		classType = new ClassType("namespace", "Application", "GtkApplication", false);
		assertFalse(classType.isClass());
		assertFalse(classType.isDirectType());

		classType = new ClassType("namespace", "Application", "GtkApplication", true);
		assertTrue(classType.isClass());
		assertTrue(classType.isDirectType());

		StructureTable.INSTANCE.add("namespace", "Application");
		classType = new ClassType("namespace", "Application", "GtkApplication*", true);
		assertTrue(classType.isClass());
		assertEquals("Application", classType.getApiTypeName("namespace"));
		assertFalse(classType.isDirectType());

		classType = new ClassType("mynamespace", "Application", "GtkApplication*", false);
		assertFalse(classType.isClass());
		assertFalse(classType.isDirectType());

		classType = new ClassType("mynamespace", "namespace.Application", "GtkApplication*", false);
		assertEquals("ch.bailu.gtk.namespace.Application", classType.getApiTypeName(""));
		assertTrue(classType.isClass());
		assertFalse(classType.isDirectType());

	}
}

File: generator/src/test/kotlin/ch/bailu/gtk/TestTypeModel.kt
package ch.bailu.gtk

import ch.bailu.gtk.model.type.CType
import ch.bailu.gtk.model.type.ClassType
import ch.bailu.gtk.model.type.JavaType
import ch.bailu.gtk.model.type.NamespaceType
import ch.bailu.gtk.table.StructureTable
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class TestTypeModel {
    @Test
    fun testCType() {
        var ctype = CType("void")
        assertFalse(ctype.isConst)
        assertFalse(ctype.isSinglePointer)
        assertEquals("void", ctype.type)
        assertFalse(ctype.contains("ApplicationFlags"))
        ctype = CType("GApplicationFlags")
        assertFalse(ctype.isConst)
        assertFalse(ctype.isSinglePointer)
        assertTrue(ctype.isDirectType)
        assertEquals("GApplicationFlags", ctype.type)
        assertTrue(ctype.contains("ApplicationFlags"))
        ctype = CType("PangoAttrShape*")
        assertFalse(ctype.isConst)
        assertTrue(ctype.isSinglePointer)
        assertFalse(ctype.isDirectType)
        assertEquals("PangoAttrShape*", ctype.type)
        ctype = CType("const gchar*")
        assertTrue(ctype.isConst)
        assertTrue(ctype.isSinglePointer)
        ctype = CType("gchar**")
        assertFalse(ctype.isConst)
        assertFalse(ctype.isSinglePointer)
    }

    @Test
    fun testJavaType() {
        var jtype = JavaType("int")
        assertEquals("int", jtype.getApiTypeName())
        assertTrue(jtype.valid)
        jtype = JavaType("void")
        assertEquals("void", jtype.getApiTypeName())
        assertTrue(jtype.valid)
        assertTrue(jtype.isVoid())
        jtype = JavaType("void")
        assertEquals("void", jtype.getApiTypeName())
        assertTrue(jtype.valid)
        assertTrue(jtype.isVoid())
        jtype = JavaType("guint8")
        assertEquals("int", jtype.getApiTypeName())
        assertTrue(jtype.valid)
        assertFalse(jtype.isVoid())
    }

    @Test
    fun testNamespaceType() {
        var type = NamespaceType("namespace", "namespace.Application")
        assertTrue(type.valid)
        assertEquals("namespace", type.namespace)
        assertEquals("Application", type.name)
        type = NamespaceType("namespace", "Application")
        assertTrue(type.valid)
        assertEquals("namespace", type.namespace)
        assertEquals("Application", type.name)
        type = NamespaceType("namespace2", "namespace.Application")
        assertTrue(type.valid)
        assertEquals("namespace", type.namespace)
        assertEquals("Application", type.name)
        val ntype = NamespaceType("GObject", "ParamSpecInt64")
        assertEquals("ParamSpecInt64", ntype.name)
        assertEquals("gobject", ntype.namespace)
        assertTrue(ntype.valid)

        assertFalse(NamespaceType.INVALID.valid)

        val geoNamespace = NamespaceType("geoclue", "Gio.AsyncReadyCallback")
        assertEquals("gio", geoNamespace.namespace)
        assertEquals("AsyncReadyCallback", geoNamespace.name)
        assertTrue(geoNamespace.valid)

        val strangeNamespace = NamespaceType("gobject", "_Value__data__union")
        assertEquals("gobject", strangeNamespace.namespace)
        assertEquals("_Value__data__union", strangeNamespace.name)
    }


    @Test
    fun testClassTypeNotInTable() {
        val classType = ClassType("namespace", "Application", "GtkApplication")
        assertFalse(classType.valid)
        assertFalse(classType.directType)
    }

    @Test
    fun testClassTypeInTable() {
        StructureTable.add("namespace", "Application", false, "")

        val classType = ClassType("namespace", "Application", "GtkApplication")
        assertTrue(classType.valid)
        assertFalse(classType.wrapper)
        assertTrue(classType.directType)

        StructureTable.clear()
    }

    @Test
    fun testClassTypInTablePointer() {
        StructureTable.add("namespace", "Application", false, "")

        var classType = ClassType("mynamespace", "Application", "GtkApplication*")
        assertFalse(classType.valid)
        assertFalse(classType.directType)

        classType = ClassType("mynamespace", "namespace.Application", "GtkApplication*")
        assertEquals("ch.bailu.gtk.namespace.Application", classType.getApiTypeName(""))
        assertTrue(classType.valid)
        assertFalse(classType.directType)

        StructureTable.clear()
    }
}
